File: tests/codes/cpp/Basic/ParametersUtils.cpp
// =======================================
//             HELPER FUNCTIONS
// =======================================

#include "ParametersUtils.h"

#include <boost/algorithm/string.hpp>
#include <memory>
#include <stdexcept>
#include "json/json.h"
#include "xmlIMAS.h"
#include <stdexcept>
#include <tuple>
#include <iostream>

std::tuple<int,double> extract_xml_values(std::string parameters)
{
    int ntimes = 0;
    double multiplication_factor = 0.0;

    std::unique_ptr<XmlImas> doc; // libxml2 doc ptr
    doc = std::unique_ptr<XmlImas>(new XmlImas);
    doc->fromMemory(parameters);

    doc->getValue("/*/ntimes", ntimes);
    doc->getValue("/*/multiplication_factor", multiplication_factor);

    return std::make_tuple(ntimes,multiplication_factor);
}

std::tuple<int,double> extract_json_values(std::string parameters)
{
    const auto rawJsonLength = static_cast<int>(parameters.length());
    JSONCPP_STRING err;
    Json::Value root;

    Json::CharReaderBuilder builder;
    const std::unique_ptr<Json::CharReader> reader(builder.newCharReader());
    if (!reader->parse(parameters.c_str(), parameters.c_str() + rawJsonLength, &root, &err))
    {
        throw std::runtime_error("Cannot open JSON parameters file");
    }

    const int ntimes = root["parameters"]["ntimes"].asInt();
    const double multiplication_factor = root["parameters"]["multiplication_factor"].asFloat();

    return std::make_tuple(ntimes,multiplication_factor);
}

std::tuple<int,double> extract_values(std::string parameters)
{
    // This is general function for ntimes and multiplication_factor parameters extraction
    boost::algorithm::trim(parameters);

    std::tuple<int,double> ntimes_multiplicationFactor = std::make_tuple(1,1.0);
    if(parameters=="")return ntimes_multiplicationFactor;

    if(parameters[0] == '<')
    {
        ntimes_multiplicationFactor = extract_xml_values(parameters);
    }
    else if(parameters[0] == '{')
    {
        ntimes_multiplicationFactor = extract_json_values(parameters);
    }
    else if(parameters[0] == '&')
    {
        std::cerr<<"Fortran namelist parameters are supported only by fortran-based actors"<<std::endl;
    }
    else
    {
        std::cerr<<"Could not recognise parameters format"<<std::endl;
    }

    return ntimes_multiplicationFactor;
}

File: tests/codes/java/Basic/ParametersUtils.java
package iwrap.test_codes.basic;

import org.w3c.dom.Document;
import org.w3c.dom.Element;
import org.w3c.dom.NodeList;

import org.xml.sax.InputSource;
import org.xml.sax.SAXException;

import org.json.JSONArray;
import org.json.JSONObject;

import javax.xml.parsers.DocumentBuilder;
import javax.xml.parsers.DocumentBuilderFactory;
import javax.xml.parsers.ParserConfigurationException;

import java.io.StringReader;
import java.io.IOException;

public class ParametersUtils
{
    // =====================================================================================
    //             HELPER FUNCTIONS
    // =====================================================================================

    public static Number[] extract_values(String parameters) throws SAXException, ParserConfigurationException, IOException
    {
        // returns [<ntimes>,<multiplication_factor>] from parameters string
        // handles xml and json parameters formats
        Number[] extracted_values = new Number[2];

        if(parameters.charAt(0) == '<')
        {
            return extract_xml_values(parameters);
        }
        else if(parameters.charAt(0) == '{')
        {
            return extract_json_values(parameters);
        }
        else if(parameters.charAt(0) == '&')
        {
           System.out.println("Fortran namelist parameters are supported only by fortran-based actors");
        }
        else
        {
            System.out.println("YAML parameters not supported yet");
        }

        return extracted_values;
    }

    public static Number[] extract_xml_values(String parameters) throws SAXException, ParserConfigurationException, IOException
    {
        // returns [<ntimes>,<multiplication_factor>] from parameters string
        Number[] extracted_values = new Number[2];

        DocumentBuilderFactory factory = DocumentBuilderFactory.newInstance();
        DocumentBuilder builder = factory.newDocumentBuilder();
        Document document = builder.parse(new InputSource(new StringReader(parameters)));
        Element rootElement = document.getDocumentElement();

        extracted_values[0] = Integer.valueOf(rootElement.getElementsByTagName("ntimes").item(0).getTextContent().trim());
        extracted_values[1] = Double.valueOf(rootElement.getElementsByTagName("multiplication_factor").item(0).getTextContent().trim());

        return extracted_values;
    }

    public static Number[] extract_json_values(String parameters)
    {
        // returns [<ntimes>,<multiplication_factor>] from parameters string
        Number[] extracted_values = new Number[2];

        JSONObject obj = new JSONObject(parameters);
        JSONObject parameters_obj = obj.getJSONObject("parameters");
        extracted_values[0] = parameters_obj.getInt("ntimes");
        extracted_values[1] = parameters_obj.getFloat("multiplication_factor");

        return extracted_values;
    }

}
